File: src/MagnusAddon/ankiutils/query_builder.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ankiutils import app
from language_services.janome_ex.word_extraction.text_analysis import TextAnalysis
from note.note_constants import Builtin, MyNoteFields, NoteFields, NoteTypes, SentenceNoteFields
from queryablecollections.collections.q_set import QSet
from sysutils import kana_utils

if TYPE_CHECKING:
    from collections.abc import Iterable

    from anki.cards import CardId
    from anki.notes import NoteId
    from note.jpnote import JPNote
    from note.kanjinote import KanjiNote
    from note.vocabulary.vocabnote import VocabNote

f_question = MyNoteFields.question
f_reading = NoteFields.Vocab.Reading
f_answer = MyNoteFields.answer
f_forms = NoteFields.Vocab.Forms
f_reading_on = NoteFields.Kanji.Reading_On
f_reading_kun = NoteFields.Kanji.Reading_Kun

card_read = f"{Builtin.Card}:{NoteFields.VocabNoteType.Card.Reading}"

note_kanji = f"{Builtin.Note}:{NoteTypes.Kanji}"
note_vocab = f"{Builtin.Note}:{NoteTypes.Vocab}"
note_sentence = f"{Builtin.Note}:{NoteTypes.Sentence}"
excluded_deck_substring = "*Excluded*"
deck_excluded = f"""{Builtin.Deck}:{excluded_deck_substring}"""

note_vocab = note_vocab

def _or_clauses(clauses: list[str]) -> str:
    return clauses[0] if len(clauses) == 1 else f"""({" OR ".join(clauses)})"""

def field_contains_word(field: str, *words: str) -> str:
    return _or_clauses([f'''"{field}:re:\\b{query}\\b"''' for query in words])

def field_contains_string(field: str, *words: str) -> str:
    return _or_clauses([f'''"{field}:*{query}*"''' for query in words])

def sentence_search(word: str, exact: bool = False) -> str:
    result = f"""{note_sentence} """

    def form_query(form: str) -> str:
        return f"""(-{field_contains_word(SentenceNoteFields.user_excluded_vocab, form)} AND ({f_question}:*{form}* OR {field_contains_word(SentenceNoteFields.parsing_result, form)} OR {field_contains_word(SentenceNoteFields.user_extra_vocab, form)}))"""

    if not exact:
        vocabs = app.col().vocab.with_form(word)
        if vocabs:
            forms: set[str] = vocabs.select_many(lambda voc: voc.forms.all_list()).to_set()  # set(ex_sequence.flatten([v.forms.all_list() for v in vocabs]))
            return result + "(" + "　OR ".join([form_query(form) for form in forms]) + ")"

    return result + f"""({form_query(word)})"""

def potentially_matching_sentences_for_vocab(word: VocabNote) -> str:
    if word.matching_configuration.requires_forbids.exact_match.is_required:
        search_strings = word.forms.all_list()
    else:
        search_strings = word.conjugator.get_stems_for_all_forms() + word.forms.all_list()
        # we'd much rather catch too much than not enough here. False positives only slow things down a bit,
        # false negatives on the other hand, can totally confuse things, giving the user entirely the wrong idea about the value of a word when no/few matches are found.
        search_strings += [form[:-1] for form in word.forms.all_list() if (len(form) > 2 or kana_utils.contains_kanji(form)) and kana_utils.character_is_kana(form[-1])]
    return f"""{note_sentence} {field_contains_string(f_question, *search_strings)}"""

def sentences_with_question_substring(substring: str) -> str:
    return f"""{note_sentence} {field_contains_string(SentenceNoteFields.active_question, substring)}"""

def notes_lookup(notes: Iterable[JPNote]) -> str:
    return notes_by_id([note.get_id() for note in notes])

def notes_by_id(note_ids: list[NoteId]) -> str:
    return f"""{NoteFields.note_id}:{",".join([str(note_id) for note_id in note_ids])}""" if note_ids else ""

def single_vocab_wildcard(query: str) -> str: return f"{note_vocab} ({f_forms}:*{query}* OR {f_reading}:*{query}* OR {f_answer}:*{query}*)"

def single_vocab_by_question_reading_or_answer_exact(query: str) -> str:
    return f"{note_vocab} ({field_contains_word(f_forms, query)} OR {field_contains_word(f_reading, kana_utils.katakana_to_hiragana(query))} OR {field_contains_word(f_answer, query)})"

def single_vocab_by_form_exact(query: str) -> str: return f"{note_vocab} {field_contains_word(f_forms, query)}"

def single_vocab_by_form_exact_read_card_only(query: str) -> str: return f"({single_vocab_by_form_exact(query)}) {card_read}"

def kanji_in_string(query: str) -> str: return f"{note_kanji} ( {' OR '.join([f'{f_question}:{char}' for char in query])} )"

def vocab_dependencies_lookup_query(vocab: VocabNote) -> str:
    def single_vocab_clause(voc: str) -> str:
        return f"{field_contains_word(f_forms, voc)}"

    def create_vocab_clause(text: str) -> str:
        dictionary_forms = TextAnalysis.from_text(text).all_words_strings()
        return f"({note_vocab} ({' OR '.join([single_vocab_clause(voc) for voc in dictionary_forms])})) OR " if dictionary_forms else ""

    def create_vocab_vocab_clause() -> str:
        return create_vocab_clause(vocab.get_question())

    def create_kanji_clause() -> str:
        return f"{note_kanji} ( {' OR '.join([f'{f_question}:{char}' for char in vocab.get_question()])} )"

    return f"""{create_vocab_vocab_clause()} ({create_kanji_clause()})"""

def vocab_with_kanji(note: KanjiNote) -> str: return f"{note_vocab} {f_forms}:*{note.get_question()}*"

def vocab_clause(voc: str) -> str:
    return f"""{field_contains_word(f_forms, voc)}"""

def text_vocab_lookup(text: str) -> str:
    dictionary_forms = TextAnalysis.from_text(text).all_words_strings()
    return vocabs_lookup(dictionary_forms)

def vocabs_lookup(dictionary_forms: list[str]) -> str:
    return f"{note_vocab} ({' OR '.join([vocab_clause(voc) for voc in dictionary_forms])})"

def vocabs_lookup_strings(words: list[str]) -> str:
    return f'''{note_vocab} ({' OR '.join([f"""{field_contains_word(f_forms, voc)}""" for voc in words])})'''

def vocabs_lookup_strings_read_card(words: list[str]) -> str:
    return f"""{vocabs_lookup_strings(words)} {card_read}"""

def kanji_with_reading_part(reading_part: str) -> str:
    hiragana_reading_part = kana_utils.anything_to_hiragana(reading_part)
    return f"""note:{NoteTypes.Kanji} ({f_reading_on}:*{hiragana_reading_part}* OR {f_reading_kun}:*{hiragana_reading_part}*)"""

def exact_matches(question: str) -> str:
    return f"""{f_question}:"{question}" OR {field_contains_word(f_forms, question)}"""

def exact_matches_no_sentences(question: str) -> str:
    return f"""({exact_matches(question)}) -{note_sentence}"""

def exact_matches_no_sentences_reading_cards(question: str) -> str:
    return f"""({exact_matches_no_sentences(question)}) {card_read} -{deck_excluded}"""

def immersion_kit_sentences() -> str:
    return f'''"{Builtin.Note}:{NoteTypes.immersion_kit}"'''

def kanji_with_radicals_in_string(search: str) -> str:
    radicals = QSet(search.strip().replace(",", "").replace(" ", ""))
    def kanji_contails_all_radicals(kanji: KanjiNote) -> bool: return not any(rad for rad in radicals if rad not in kanji.get_radicals())
    return (radicals
            .select_many(lambda radical: app.col().kanji.with_radical(radical))
            .where(kanji_contails_all_radicals)
            .pipe(notes_lookup))

def open_card_by_id(card_id: CardId) -> str:
    return f"cid:{card_id}"

def kanji_with_meaning(search: str) -> str:
    return f"""{note_kanji} ({f_answer}:*{search}*)"""

File: src/src_dotnet/JAStudio.Core/AnkiUtils/QueryBuilder.cs
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.LanguageServices;
using JAStudio.Core.LanguageServices.JanomeEx;
using JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;
using JAStudio.Core.Note;
using JAStudio.Core.Note.Collection;

namespace JAStudio.Core.AnkiUtils;

/// <summary>
/// Builds Anki search query strings for finding notes and cards.
/// Ported from jastudio/ankiutils/query_builder.py
/// </summary>
public class QueryBuilder
{
   readonly VocabCollection _vocab;
   readonly KanjiCollection _kanji;
   readonly AnalysisServices _analysisServices;
   readonly AnkiNoteIdMap _ankiNoteIdMap;

   internal QueryBuilder(VocabCollection vocab, KanjiCollection kanji, AnalysisServices analysisServices, AnkiNoteIdMap ankiNoteIdMap)
   {
      _vocab = vocab;
      _kanji = kanji;
      _analysisServices = analysisServices;
      _ankiNoteIdMap = ankiNoteIdMap;
   }

   const string ExcludedDeckSubstring = "*Excluded*";

    /// <summary>
    /// Combines clauses with OR, wrapping in parentheses if multiple clauses.
    /// </summary>
    string OrClauses(IEnumerable<string> clauses)
    {
        var clauseList = clauses.ToList();
        return clauseList.Count == 1 ? clauseList[0] : $"({string.Join(" OR ", clauseList)})";
    }

    /// <summary>
    /// Creates a field search for whole word matches using regex word boundaries.
    /// </summary>
    string FieldContainsWord(string field, params string[] words)
    {
        return OrClauses(words.Select(word => $"\"{field}:re:\\\\b{word}\\\\b\""));
    }

    /// <summary>
    /// Creates a field search for substring matches using wildcards.
    /// </summary>
    string FieldContainsString(string field, params string[] words)
    {
        return OrClauses(words.Select(word => $"\"{field}:*{word}*\""));
    }

    /// <summary>
    /// Search for sentence notes containing the given word.
    /// If exact=false, searches for all forms of vocab notes matching the word.
    /// Ported from sentence_search()
    /// </summary>
    public string SentenceSearch(string word, bool exact = false)
    {
        var result = $"{Builtin.Note}:{NoteTypes.Sentence} ";

        string FormQuery(string form) =>
            $"(({MyNoteFields.Question}:*{form}* OR {FieldContainsWord(SentenceNoteFields.ParsingResult, form)}))";

        if (!exact)
        {
            var vocabs = _vocab.WithForm(word).ToList();
            if (vocabs.Any())
            {
                var forms = vocabs.SelectMany(voc => voc.Forms.AllList()).Distinct().ToList();
                return result + "(" + string.Join("　OR ", forms.Select(FormQuery)) + ")";
            }
        }

        return result + $"({FormQuery(word)})";
    }

    /// <summary>
    /// Search for sentences that potentially match a vocab word.
    /// Ported from potentially_matching_sentences_for_vocab()
    /// </summary>
    public string PotentiallyMatchingSentencesForVocab(VocabNote vocab)
    {
        var searchStrings = vocab.MatchingConfiguration.RequiresForbids.Surface.IsRequired
            ? vocab.Forms.AllList().ToList()
            : vocab.Conjugator.GetStemsForAllForms().Concat(vocab.Forms.AllList()).ToList();

        return $"{Builtin.Note}:{NoteTypes.Sentence} {FieldContainsString(MyNoteFields.Question, searchStrings.ToArray())}";
    }

    /// <summary>
    /// Creates a query to find notes by their JPNote instances.
    /// Ported from notes_lookup()
    /// </summary>
    public string NotesLookup(IEnumerable<JPNote> notes)
    {
        return NotesByIds(notes.Select(n => n.GetId()));
    }

    /// <summary>
    /// Creates a query to find notes by their domain NoteIds.
    /// Converts to Anki long IDs for the query string.
    /// </summary>
    public string NotesByIds(IEnumerable<NoteId> noteIds)
    {
        var ankiIds = noteIds
            .Select(id => _ankiNoteIdMap.ToAnkiId(id))
            .Where(id => id.HasValue)
            .Select(id => id!.Value)
            .ToList();
        return ankiIds.Count > 0 ? $"{NoteFieldsConstants.NoteId}:{string.Join(",", ankiIds)}" : "";
    }

    /// <summary>
    /// Search for vocab notes using wildcard matching on forms, reading, or answer.
    /// Ported from single_vocab_wildcard()
    /// </summary>
    public string SingleVocabWildcard(string form) => $"{Builtin.Note}:{NoteTypes.Vocab} ({NoteFieldsConstants.Vocab.Forms}:*{form}* OR {NoteFieldsConstants.Vocab.Reading}:*{form}* OR {MyNoteFields.Answer}:*{form}*)";

    /// <summary>
    /// Search for vocab notes by exact word match in forms, reading, or answer.
    /// Ported from single_vocab_by_question_reading_or_answer_exact()
    /// </summary>
    public string SingleVocabByQuestionReadingOrAnswerExact(string search)
    {
        var hiraganaSearch = KanaUtils.KatakanaToHiragana(search);
        return $"{Builtin.Note}:{NoteTypes.Vocab} ({FieldContainsWord(NoteFieldsConstants.Vocab.Forms, search)} OR {FieldContainsWord(NoteFieldsConstants.Vocab.Reading, hiraganaSearch)} OR {FieldContainsWord(MyNoteFields.Answer, search)})";
    }

    /// <summary>
    /// Search for vocab notes by exact form match.
    /// Ported from single_vocab_by_form_exact()
    /// </summary>
    public string SingleVocabByFormExact(string form) => $"{Builtin.Note}:{NoteTypes.Vocab} {FieldContainsWord(NoteFieldsConstants.Vocab.Forms, form)}";

    /// <summary>
    /// Search for vocab reading cards by exact form match.
    /// Ported from single_vocab_by_form_exact_read_card_only()
    /// </summary>
    public string SingleVocabByFormExactReadCardOnly(string form) => $"({SingleVocabByFormExact(form)}) {Builtin.Card}:{NoteFieldsConstants.VocabNoteType.Card.Reading}";

    /// <summary>
    /// Search for kanji notes for all kanji characters in the string.
    /// Ported from kanji_in_string()
    /// </summary>
    public string KanjiInString(string text)
    {
        var kanjiClauses = text.Select(c => $"{MyNoteFields.Question}:{c}");
        return $"{Builtin.Note}:{NoteTypes.Kanji} ( {string.Join(" OR ", kanjiClauses)} )";
    }

    /// <summary>
    /// Search for vocab notes containing a specific kanji.
    /// Ported from vocab_with_kanji()
    /// </summary>
    public string VocabWithKanji(KanjiNote kanji) => $"{Builtin.Note}:{NoteTypes.Vocab} {NoteFieldsConstants.Vocab.Forms}:*{kanji.GetQuestion()}*";

    /// <summary>
    /// Search for vocab notes by dictionary forms extracted from text.
    /// Ported from text_vocab_lookup()
    /// </summary>
    public string TextVocabLookup(string text)
    {
        var dictionaryForms = TextAnalysis.FromText(_analysisServices, text).AllWordsStrings();
        return VocabsLookup(dictionaryForms);
    }

    /// <summary>
    /// Creates a vocab clause for a single form.
    /// Ported from vocab_clause()
    /// </summary>
    string VocabClause(string form) => FieldContainsWord(NoteFieldsConstants.Vocab.Forms, form);

    /// <summary>
    /// Search for vocab notes matching any of the given dictionary forms.
    /// Ported from vocabs_lookup()
    /// </summary>
    public string VocabsLookup(IEnumerable<string> dictionaryForms)
    {
        var forms = dictionaryForms.ToList();
        var clauses = forms.Select(VocabClause);
        return $"{Builtin.Note}:{NoteTypes.Vocab} ({string.Join(" OR ", clauses)})";
    }

    /// <summary>
    /// Search for vocab notes matching any of the given words (exact word match).
    /// Ported from vocabs_lookup_strings()
    /// </summary>
    public string VocabsLookupStrings(IEnumerable<string> words)
    {
        var wordList = words.ToList();
        var clauses = wordList.Select(word => FieldContainsWord(NoteFieldsConstants.Vocab.Forms, word));
        return $"{Builtin.Note}:{NoteTypes.Vocab} ({string.Join(" OR ", clauses)})";
    }

    /// <summary>
    /// Search for vocab reading cards matching any of the given words.
    /// Ported from vocabs_lookup_strings_read_card()
    /// </summary>
    public string VocabsLookupStringsReadCard(IEnumerable<string> words) => $"{VocabsLookupStrings(words)} {Builtin.Card}:{NoteFieldsConstants.VocabNoteType.Card.Reading}";

    /// <summary>
    /// Search for kanji notes whose readings contain the given reading part.
    /// Ported from kanji_with_reading_part()
    /// </summary>
    public string KanjiWithReadingPart(string readingPart)
    {
        var hiraganaReading = KanaUtils.AnythingToHiragana(readingPart);
        return $"{Builtin.Note}:{NoteTypes.Kanji} ({NoteFieldsConstants.Kanji.ReadingOn}:*{hiraganaReading}* OR {NoteFieldsConstants.Kanji.ReadingKun}:*{hiraganaReading}*)";
    }

    /// <summary>
    /// Search for notes with exact question match or form match.
    /// Ported from exact_matches()
    /// </summary>
    public string ExactMatches(string question) => $"{MyNoteFields.Question}:\"{question}\" OR {FieldContainsWord(NoteFieldsConstants.Vocab.Forms, question)}";

    /// <summary>
    /// Search for exact matches excluding sentence notes.
    /// Ported from exact_matches_no_sentences()
    /// </summary>
    public string ExactMatchesNoSentences(string question) => $"({ExactMatches(question)}) -{Builtin.Note}:{NoteTypes.Sentence}";

    /// <summary>
    /// Search for exact matches on reading cards, excluding sentences and excluded decks.
    /// Ported from exact_matches_no_sentences_reading_cards()
    /// </summary>
    public string ExactMatchesNoSentencesReadingCards(string question) => $"({ExactMatchesNoSentences(question)}) {Builtin.Card}:{NoteFieldsConstants.VocabNoteType.Card.Reading} -{Builtin.Deck}:{ExcludedDeckSubstring}";

    /// <summary>
    /// Search for Immersion Kit sentence notes.
    /// Ported from immersion_kit_sentences()
    /// </summary>
    public string ImmersionKitSentences() => $"\"{Builtin.Note}:{NoteTypes.ImmersionKit}\"";

    /// <summary>
    /// Search for kanji notes containing all specified radicals.
    /// Ported from kanji_with_radicals_in_string()
    /// </summary>
    public string KanjiWithRadicalsInString(string search)
    {
        var radicals = search.Trim().Replace(",", "").Replace(" ", "").ToCharArray().ToHashSet();

        bool KanjiContainsAllRadicals(KanjiNote kanji)
        {
            var kanjiRadicals = kanji.Radicals.ToHashSet();
            return radicals.All(r => kanjiRadicals.Contains(r.ToString()));
        }

        var matchingKanji = radicals
            .SelectMany(radical => _kanji.WithRadical(radical.ToString()))
            .Distinct()
            .Where(KanjiContainsAllRadicals)
            .ToList();

        return NotesLookup(matchingKanji);
    }

    /// <summary>
    /// Search for a card by its ID.
    /// Ported from open_card_by_id()
    /// </summary>
    public string OpenCardById(long cardId) => $"cid:{cardId}";

    /// <summary>
    /// Search for kanji notes whose meaning contains the search string.
    /// Ported from kanji_with_meaning()
    /// </summary>
    public string KanjiWithMeaning(string search) => $"{Builtin.Note}:{NoteTypes.Kanji} ({MyNoteFields.Answer}:*{search}*)";

    /// <summary>
    /// Search for vocab dependencies (vocab and kanji that a word depends on).
    /// Ported from vocab_dependencies_lookup_query()
    /// </summary>
    public string VocabDependenciesLookupQuery(VocabNote vocab)
    {
        string CreateVocabClause(string text)
        {
            var dictionaryForms = TextAnalysis.FromText(_analysisServices, text).AllWordsStrings();
            if (!dictionaryForms.Any())
                return "";

            var clauses = dictionaryForms.Select(VocabClause);
            return $"({Builtin.Note}:{NoteTypes.Vocab} ({string.Join(" OR ", clauses)})) OR ";
        }

        string CreateKanjiClause()
        {
            var kanjiClauses = vocab.GetQuestion().Select(c => $"{MyNoteFields.Question}:{c}");
            return $"{Builtin.Note}:{NoteTypes.Kanji} ( {string.Join(" OR ", kanjiClauses)} )";
        }

        var vocabVocabClause = CreateVocabClause(vocab.GetQuestion());
        var kanjiClause = CreateKanjiClause();

        return $"{vocabVocabClause}({kanjiClause})";
    }
}
